components/servo_tracing/lib.rs

```diff
@@ -0,0 +1,394 @@
+use quote::{ToTokens, TokenStreamExt, quote};
+use syn::parse::{Parse, Parser};
+use syn::punctuated::Punctuated;
+use syn::token::Comma;
+use syn::{Expr, ItemFn, Meta, MetaList, Token, parse_quote, parse2};
+
+struct Fields(MetaList);
+impl From<MetaList> for Fields {
+    fn from(value: MetaList) -> Self {
+        Fields(value)
+    }
+}
+
+impl Fields {
+    fn inject_servo_profiling(&mut self) -> syn::Result<()> {
+        let mut servo_profiling_given = false;
+        let metalist = std::mem::replace(&mut self.0, parse_quote! {fields()});
+
+        let arguments: Punctuated<Meta, Comma> =
+            Punctuated::parse_terminated.parse2(metalist.tokens)?;
+        for arg in arguments.iter() {
+            servo_profiling_given |= arg.path().is_ident("servo_profiling");
+        }
+
+        let metalist = if servo_profiling_given {
+            parse_quote! {
+                fields(#arguments)
+            }
+        } else {
+            parse_quote! {
+                fields(servo_profiling=true, #arguments)
+            }
+        };
+
+        let _ = std::mem::replace(&mut self.0, metalist);
+
+        Ok(())
+    }
+}
+
+impl ToTokens for Fields {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        let items = &self.0;
+        tokens.append_all(quote! { #items });
+    }
+}
+enum Directive {
+    Passthrough(Meta),
+    Level(Expr),
+    Fields(Fields),
+}
+
+impl Directive {
+    fn is_fields(&self) -> bool {
+        matches!(self, Directive::Fields(..))
+    }
+
+    fn is_level(&self) -> bool {
+        matches!(self, Directive::Level(..))
+    }
+
+    fn fields_mut(&mut self) -> Option<&mut Fields> {
+        match self {
+            Directive::Fields(fields) => Some(fields),
+            _ => None,
+        }
+    }
+}
+
+impl ToTokens for Directive {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        match self {
+            Directive::Passthrough(meta) => tokens.append_all(quote! { #meta }),
+            Directive::Level(level) => tokens.append_all(quote! { level = #level }),
+            Directive::Fields(fields) => tokens.append_all(quote! { #fields }),
+        };
+    }
+}
+
+impl ToTokens for InstrumentConfiguration {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        tokens.append_terminated(&self.0, Punct::new(',', proc_macro2::Spacing::Joint));
+    }
+}
+
+struct InstrumentConfiguration(Vec<Directive>);
+
+impl InstrumentConfiguration {
+    fn inject_servo_profiling(&mut self) -> syn::Result<()> {
+        // establish invariant that self.0 has a Directive::Fields.
+
+        let contains = self.0.iter().any(|a| a.is_fields());
+        if !contains {
+            self.0.push(Directive::Fields(Fields(
+                parse_quote! { fields(servo_profiling = true) },
+            )));
+        }
+        let component = self.0.iter_mut().find(|a| a.is_fields()).unwrap();
+
+        assert!(matches!(component, Directive::Fields(..)));
```

Comment:
Does this assert add anything, given the panic in the branch below if the condition doesn't hold?

Reply:
I find that it adds some clarity / is more explicit over what we expect; whereas the other one is implicit / one needs to lookup the implementation, though is minor and we can remove it.